src/Stratis.Bitcoin.Features.Consensus/CoinViews/CachedCoinView.cs

```diff
@@ -160,25 +169,97 @@ public CachedCoinView(Network network, ICheckpoints checkpoints, ICoindb coindb,
+                    return;
+
+                Flush();
+
+                ChainedHeader fork = chainIndexer[coinViewTip.Hash];
+                if (fork == null)
+                {
+                    // Determine the last usable height.
+                    int height = BinarySearch.BinaryFindFirst(h => (h > chainIndexer.Height) || this.GetRewindData(h).PreviousBlockHash.Hash != chainIndexer[h].Previous.HashBlock, 0, coinViewTip.Height + 1) - 1;
+                    fork = chainIndexer[(height < 0) ? coinViewTip.Height : height];
+                }
+
+                while (coinViewTip.Height != fork.Height)
+                {
+                    if ((coinViewTip.Height % 100) == 0)
+                        this.logger.LogInformation("Rewinding coin view from '{0}' to {1}.", coinViewTip, fork);
+
+                    // If the block store was initialized behind the coin view's tip, rewind it to on or before it's tip.
+                    // The node will complete loading before connecting to peers so the chain will never know that a reorg happened.
+                    coinViewTip = this.coindb.Rewind(new HashHeightPair(fork));
+                };
+            }
+        }
+
+        public void Initialize(ChainedHeader chainTip, ChainIndexer chainIndexer, ConsensusRulesContainer consensusRulesContainer)
         {
             this.coindb.Initialize(chainTip);
 
+            Sync(chainIndexer);
+
             HashHeightPair coinViewTip = this.coindb.GetTipHash();
 
-            while (true)
+            // If the coin view is behind the block store then catch up from the block store.
+            if (coinViewTip.Height < chainTip.Height)
             {
-                ChainedHeader pendingTip = chainTip.FindAncestorOrSelf(coinViewTip.Hash);
+                try
+                {
+                    var loadCoinViewRule = consensusRulesContainer.FullValidationRules.OfType<LoadCoinviewRule>().Single();
+                    var saveCoinViewRule = consensusRulesContainer.FullValidationRules.OfType<SaveCoinviewRule>().Single();
+                    var coinViewRule = consensusRulesContainer.FullValidationRules.OfType<CoinViewRule>().Single();
+                    var deploymentsRule = consensusRulesContainer.FullValidationRules.OfType<SetActivationDeploymentsFullValidationRule>().Single();
+
+                    foreach ((ChainedHeader chainedHeader, Block block) in this.blockStore.BatchBlocksFrom(chainIndexer[coinViewTip.Hash], chainIndexer, this.cancellationToken, batchSize: 1000))
+                    {
+                        if (block == null)
+                            break;
+
+                        if ((chainedHeader.Height % 10000) == 0)
+                        {
+                            this.Flush(true);
+                            this.logger.LogInformation("Rebuilding coin view from '{0}' to {1}.", chainedHeader, chainTip);
+                        }
+
+                        var utxoRuleContext = new PosRuleContext()
```

Comment:
Just wondering, Cirrus would be using a subclass of the `PowRuleContext` would it not? So maybe we should try to get the correct type of context from the rule engine

Reply:
Fixed.